File: src/main/java/editor/filter/FaceSearchOptions.java
package editor.filter;

import java.awt.Image;
import java.io.IOException;

import javax.imageio.ImageIO;
import javax.swing.Icon;
import javax.swing.ImageIcon;

/**
 * This enumerats all the ways that a user can select to filter faces when filtering cards: any face, all faces,
 * front face only, or back face only.  When a card has more than two faces (like Who/What/When/Where/Why), it isn't
 * possible to only filter using middle faces.  This doesn't affect cards that have only one face.
 */
public enum FaceSearchOptions
{
    /** Any face of a card can match the filter. */
    ANY,
    /** All faces of a card must match the filter. */
    ALL,
    /** Front (or only) face of a card must match the filter. */
    FRONT,
    /** Back (or only) face of a card must match the filter. */
    BACK;
    
    /** Icon to display on a filter line to indicate which face to search for that line. */
    private final ImageIcon icon;

    private FaceSearchOptions()
    {
        ImageIcon img;
        try
        {
            img = new ImageIcon(ImageIO.read(getClass().getResourceAsStream("/images/faces/" + toString().toLowerCase() + ".png")));
        }
        catch (IOException e)
        {
            img = new ImageIcon();
            e.printStackTrace();
        }
        icon = img;
    }

    /**
     * @return The icon corresponding to this FacesFilter at its original size.
     */
    public Icon getIcon()
    {
        return icon;
    }

    /**
     * Scale this FacesFilter's icon.
     * 
     * @param width new width of the icon
     * @return The icon scaled so that its width is the specified size.
     */
    public Icon getIcon(int width)
    {
        return new ImageIcon(icon.getImage().getScaledInstance(width, -1, Image.SCALE_SMOOTH));
    }
}

File: src/main/scala/editor/filter/FaceSearchOptions.scala
package editor.filter

import java.awt.Image
import java.io.IOException
import javax.imageio.ImageIO
import javax.swing.ImageIcon

/**
 * All the ways a filter can be applied to the faces of a multi-faced card. Has no effect on single-
 * faced cards.
 * 
 * @constructor create a new way to apply a filter to a card's faces
 * @param tooltip hint on what the symbol used to represent the option means
 * 
 * @author Alec Roelke
 */
enum FaceSearchOptions(val tooltip: String) {
  /** Icon to display to indicate the current way to filter faces for a filter. */
  lazy val icon = try {
    ImageIcon(ImageIO.read(getClass.getResourceAsStream(s"/images/faces/${toString.toLowerCase}.png")))
  } catch case e: IOException => {
    e.printStackTrace
    ImageIcon()
  }

  /**
   * Get a resized version of the icon for a filtering option.
   * 
   * @param width new side size of the icon
   * @return a scaled instance of the option's icon
   */
  def scaled(width: Int) = ImageIcon(icon.getImage.getScaledInstance(width, -1, Image.SCALE_SMOOTH))

  /** A card passes the filter if any of its faces pass. */
  case ANY   extends FaceSearchOptions("Can match any face")
  /** A card only passes the filter if all of its faces pass. */
  case ALL   extends FaceSearchOptions("Must match all faces")
  /** A card passes the filter if its front face passes, regardless of whether or not the back does. */
  case FRONT extends FaceSearchOptions("Only has to match the front face")
  /** A card passes the filter if its back face passes, regardless of whether or not the front does. */
  case BACK  extends FaceSearchOptions("Only has to match the back face")
}
